Added support for binding the raw request body

davidfowl opened the pull request:
- Added support for PipeReader, Stream, ~ReadOnlySequence<byte>, ReadOnlyMemory<byte> and byte[]~
- Added test

~This will potentially break custom JSON converters for byte[] (maybe base64 JSON encoded content?)~

Fixes https://github.com/dotnet/aspnetcore/issues/38153

~TBD: Fix the error messages that talk about JSON binding when these types are being used.~

davidfowl committed:
Added support for binding the raw request body
- Added support for Stream, ReadOnlySequence<byte>, ReadOnlyMemory<byte> and byte[]
- Added test

JamesNK commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -679,6 +691,37 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
+                        {
+                            if (bodyType == typeof(ReadOnlySequence<byte>))
+                            {
+                                // REVIEW: Does this need to be a copy? We can tell users to consume the buffer
+                                // immediately in the action
+                                return (buffer, true);
```
When is `PipeReader.AdvanceTo` called to say the body is consumed? Before or after the request delegate is executed?

davidfowl replied:
After thinking about this, it's actually broken for non Kestrel servers since nobody consumes that pipe currently. I'll need to add more logic there.

davidfowl replied:
Hmm actually no, it seems like we complete the pipe reader in both cases https://github.com/dotnet/aspnetcore/blob/694507a5d24fd858dff8231829e42fb21acbe9b9/src/Http/Http/src/Features/RequestBodyPipeFeature.cs#L35-L48

JamesNK commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -679,6 +691,37 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
+                            // LOH be damned
+                            if (bodyType == typeof(ReadOnlyMemory<byte>))
+                            {
+                                return (new ReadOnlyMemory<byte>(buffer.ToArray()), true);
```
Assuming that not copying is correct for the entire `ReadOnlySequence` above, then here you could check to see whether the ROS is a single segment and if it is then return `buffer.First`

davidfowl replied:
I'm not sure if we want to support this but I figured I'd put it in the PR and have a discussion. Copying is the safest thing to do here but I'm trying to figure out if there's a type we can use to indicate that a copy shouldn't be made (ReadOnlySequence\<byte\>).

davidfowl replied:
Decided to only support ReadOnlySequence\<byte\>

davidfowl committed:
Fix some issues add more features
- Advance the PipeReader preemptively in the cases where we copy the buffer.
- Add support for PipeReader as an argument.

davidfowl committed:
Random clean up

davidfowl commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -679,6 +697,43 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
+                            }
+
+                            // This is very LOH unfriendly
+                            var copiedBuffer = buffer.ToArray();
+
+                            // Consume the whole thing
+                            bodyReader.AdvanceTo(buffer.End);
+
+                            if (bodyType == typeof(ReadOnlyMemory<byte>))
+                            {
+                                return (new ReadOnlyMemory<byte>(copiedBuffer), true);
+                            }
+
+                            return (copiedBuffer, true);
+                        }
+
+                        // Buffer the body
+                        bodyReader.AdvanceTo(buffer.Start, buffer.End);
```
These are up for debate. It's fine for small buffers but not huge ones. We could only support `ReadOnlySequence<byte>` in a non-copying way and then force users to call `ToArray()` in their code. Another option would be to use the array pool on their behalf but I don't like that for arrays or memory.

davidfowl replied:
I decided to only support ReadOnlySequence\<byte\>

davidfowl committed:
Remove support for byte[] and ReadOnlyMemory<byte>

davidfowl committed:
Fix method names and remove extra fluff

davidfowl committed:
Removed extra space

davidfowl committed:
Removed JSON from the error message

davidfowl committed:
Formatting

davidfowl commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -679,6 +695,25 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
+
+                        if (result.IsCompleted)
+                        {
+                            return (buffer, true);
```
This ends up boxing the ReadOnlySequence\<byte\> unfortunately 😢

davidfowl committed:
Remove the boxing for raw body binding
- Split the code path for raw body binding with ReadOnlySequence<byte> to avoid boxing.

davidfowl committed:
Added more tests

davidfowl commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -582,19 +596,105 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
+            };
+        }
+
+        static async Task<(ReadOnlySequence<byte>, bool)> TryReadBodyAsync(HttpContext httpContext)
```
We should get rid of the extra task allocation and state machine here.

davidfowl committed:
Small nit

davidfowl committed:
Added more tests

martincostello commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -582,19 +596,102 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
+                }
+            }
+
+            return (default, true);
```
Did you mean to return false here instead? Otherwise this method only returns true for the flag so isn't really a `Try...()` and could just return the Task<ROS>.

davidfowl replied:
👍🏾

davidfowl committed:
Remove tuple and use ValueTask instead of Task

halter73 commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -582,19 +596,92 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
+
+                    if (result.IsCompleted)
+                    {
+                        return buffer;
```
I think calling the route handler with the reader in an incomplete/unadvanced state is problematic. Any code accessing the `BodyReader`, either through the `HttpContext` or as a separate `PipeReader` parameter, will be unable to partially examine the body. [Any attempt will throw an InvalidOperationException](https://github.com/dotnet/runtime/blob/32a24b8b1fdf45dde7582ceb8812559e07e8454b/src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs#L468-L471) complaining that "The examined position cannot be less than the previously examined position."

I think we're better off just fully consuming the body if you take a `ReadOnlySequence<byte>` parameter.

davidfowl replied:
I don’t want to copy but I appreciate the potential problems. I’d rather leave that with the caveats or cut the whole feature (which would make me sad but I’d get over it).

halter73 replied:
Given this endpoint:

```C#
app.MapPost("/read-twice", async (HttpContext httpContext, ReadOnlySequence<byte> body) =>
{
    await httpContext.Request.Body.ReadAsync(new byte[1], 0, 0);
});
```

you get the following error:

```
System.InvalidOperationException: Reading is already in progress.
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1ContentLengthMessageBody.ReadAsyncInternal(CancellationToken cancellationToken) in C:\dev\dotnet\aspnetcore\src\Servers\Kestrel\Core\src\Internal\Http\Http1ContentLengthMessageBody.cs:line 34
   at System.Runtime.CompilerServices.PoolingAsyncValueTaskMethodBuilder`1.StateMachineBox`1.System.Threading.Tasks.Sources.IValueTaskSource<TResult>.GetResult(Int16 token)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream.ReadAsyncInternal(Memory`1 destination, CancellationToken cancellationToken) in C:\dev\dotnet\aspnetcore\src\Servers\Kestrel\Core\src\Internal\Http\HttpRequestStream.cs:line 114
   at Program.<>c.<<<Main>$>b__0_7>d.MoveNext() in C:\dev\dotnet\aspnetcore\src\Http\samples\MinimalSample\Program.cs:line 42
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Http.RequestDelegateFactory.<>c__DisplayClass54_1.<<HandleRequestBodyAndCompileRequestDelegateForRawBody>b__2>d.MoveNext() in C:\dev\dotnet\aspnetcore\src\Http\Http.Extensions\src\RequestDelegateFactory.cs:line 647

```

I don't think this is acceptable. We should either:

1. Consume the entire body so this read returns 0
2. Buffer the entire body so you can actually read it again
3. Poison the body so that it throws a better error than `System.InvalidOperationException: Reading is already in progress.`
4. Cut the whole feature

davidfowl replied:
OK I [pushed another commit](https://github.com/dotnet/aspnetcore/pull/39388/commits/df767a76a7794d9a9d868f3ca3659d223d4d278d) with option 1a. I set the body to Stream.Null. This fixes trying to consume the body after it's already been consumed by our binding logic.

pranavkm commented on src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs:
```diff
@@ -1376,6 +1377,82 @@ public async Task RequestDelegatePopulatesFromBodyParameter(Delegate action)
+                httpContext.Items.Add("body", body.ToArray());
+            }
+
+            void TestStream(HttpContext httpContext, Stream stream)
```
Neat.

pranavkm commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -1445,8 +1551,9 @@ private class FactoryContext
-        public ParameterInfo? JsonRequestBodyParameter { get; set; }
+        public ParameterInfo? RequestBodyParameter { get; set; }
         public bool AllowEmptyRequestBody { get; set; }
+        public bool IsRawRequestBody => RequestBodyParameter?.ParameterType == typeof(ReadOnlySequence<byte>);
```
It'd be interesting to see if people actually start using it. It's a somewhat obtuse type to actually use (particularly when you can bind the PipeReader instead).

davidfowl replied:
It’s something we’d have to advertise. I had support for ReadOnlyMemory\<byte\> and byte[] but removed them.

davidfowl commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -582,19 +596,92 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
+        {
+            var feature = httpContext.Features.Get<IHttpRequestBodyDetectionFeature>();
+
+            if (feature?.CanHaveBody == true)
```
If allow empty body is true it’ll bind otherwise 400

davidfowl commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -582,19 +596,92 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
+        if (factoryContext.ParameterBinders.Count > 0)
+        {
+            // We need to generate the code for reading from the body before calling into the delegate
+            var continuation = Expression.Lambda<Func<object?, HttpContext, ReadOnlySequence<byte>, object?[], Task>>(
```
This could be either but using Task vs ValueTask but there’s no benefit to using ValueTask here. There’s no pooling to be had and synchronous completion won’t allocate regardless

davidfowl commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -582,19 +596,92 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
+
+            // Looping over arrays is faster
+            var binders = factoryContext.ParameterBinders.ToArray();
+            var count = binders.Length;
```
I originally didn’t do it would only capture these locals only (in case we nulled out the factory context later).

davidfowl committed:
Set the body to a null stream after consuming it
- This prevents code that would otherwise read the body from doing so after consuming it.

Tratcher commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -582,19 +596,96 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
+
+                    if (result.IsCompleted)
+                    {
+                        // We're not buffering the body so we want to block consuming code from reading again
+                        // and getting weird errors. Treat further reads as a fully consumed body.
+                        httpContext.Request.Body = Stream.Null;
```
This is odd, what if someone else is buffering the body?

pranavkm replied:
I guess you'd have to check if the body is seekeable, if so seek it. RadFormAsync / MVC doesn't change / reset the Body once it has been read, couldn't we do the same here?

halter73 replied:
I think seeking given a seekable stream is fine.

If we're replacing the body anyway, we should advance the reader before returning even if Kestrel will do it anyway.

We would have to do this after invoking the route handler, so memory doesn't get returned too soon. I'm still worried about people keeping a reference to the `ReadOnlySequence<byte>` after the route handler completes if we don't copy.

davidfowl replied:
> If we're replacing the body anyway, we should advance the reader before returning even if Kestrel will do it anyway.

Just for cleanliness? It's not just Kestrel doing it, it would happen for any properly implemented pipereader (it will complete the pipereader).

> I'm still worried about people keeping a reference to the ReadOnlySequence<byte> after the route handler completes if we don't copy.

Yea but this feature is about not copying the buffer. That's why it's a `ROS<byte>`. That's the one thing I'm unwilling to compromise here on as I think it makes the original scenario (posted in the issue) less efficient.

davidfowl replied:
> This is odd, what if someone else is buffering the body?

I don't see how that affects this? I can revert and set the the original stream after executing the handler. The body is being buffered for the logic that is consuming the buffer.

halter73 replied:
> Just for cleanliness? It's not just Kestrel doing it, it would happen for any properly implemented pipereader (it will complete the pipereader).

There's no guarantee the next call will be to `Complete()`. It could matter if someone writes request body draining middleware that waits a different amount of time than Kestrel's 5 seconds. Basically, why not? Especially if we end up doing Stream resetting logic after the route handler completes anyway.

davidfowl replied:
We have the reference to the pipe that we read and we can complete it. It's just redundant is all.

halter73 replied:
It's more about generic code that might try to read any unconsumed bytes from the body after the route handler runs but before middleware completes.

davidfowl replied:
Also @Tratcher I don't think we should check is seekable here. Specifically because we need to poison both the stream and the PipeReader to be unusable 😄

Tratcher replied:
At least revert it when you're done then.

davidfowl replied:
Done.

davidfowl committed:
PR feedback
- Restore the original stream after the handler executes.

davidfowl commented on src/Http/Http.Extensions/src/RequestDelegateFactory.cs:
```diff
@@ -582,19 +596,92 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
+
+    private static Func<object?, HttpContext, Task> HandleRequestBodyAndCompileRequestDelegateForRawBody(Expression responseWritingMethodCall, FactoryContext factoryContext)
+    {
+        Debug.Assert(factoryContext.RequestBodyParameter is not null, "factoryContext.RequestBodyParameter is null for a body parameter.");
```
Don't think so.

davidfowl committed:
Make sure we can read from the original pipereader when we unwind from the call.

davidfowl committed:
Support scenarios where nulling out the stream doesn't change the pipereader

davidfowl committed:
Remove support for the raw body as a ROS<byte>

davidfowl committed:
Small reverts

davidfowl committed:
Fixed more things

davidfowl committed:
Moar

davidfowl committed:
Reverted more stuff

davidfowl committed:
One more I hope

davidfowl committed:
Fixed the test

davidfowl commented:
This PR is ready now. I removed reading the request body as a raw buffer and am only supporting Stream and PipeReader.

BrennanConroy reviewed (approved):
Looks a lot simpler than before 😃 

Do we have an issue to track doc updates for .NET 7? We need to add IFormFileCollection, PipeReader, and Stream to https://docs.microsoft.com/aspnet/core/fundamentals/minimal-apis?view=aspnetcore-6.0#special-types now

davidfowl commented:
> Do we have an issue to track doc updates for .NET 7? We need to add IFormFileCollection, PipeReader, and Stream to https://docs.microsoft.com/aspnet/core/fundamentals/minimal-apis?view=aspnetcore-6.0#special-types now

We don't have one yet. Good idea.
